### ISSUE TITLE
osx_c.sh build script fails with "no rule to make target /usr/lib/x86_64-linux-gnu/libssl.so"

### ISSUE THREAD

**Issue by kiranpradeep** [Azure/iot-edge#281]
For building the SDK on mac os x 10.12.3, I was using `osx_c.sh` script and faced the following issue.

> make[2]: *** No rule to make target `/usr/lib/x86_64-linux-gnu/libssl.so', needed by core/libgateway.dylib.  Stop.

I had explicity set OPENSSL_ROOT_DIR to custom directory, still iot-edge continues to look for libssl in /usr/lib/x86_64-linux-gnu/.
Terminal copied below

> machine_name:jenkins user_name$ pwd
> /Users/user_name/src/try/azure-iot-gateway-sdk/jenkins
> machine_name:jenkins user_name$ 
> machine_name:jenkins user_name$ export OPENSSL_ROOT_DIR=/Users/user_name/base/openssl 
> machine_name:jenkins user_name$ 
> machine_name:jenkins user_name$ ./osx_c.sh 
> ~/src/try/azure-iot-gateway-sdk/build ~/src/try/azure-iot-gateway-sdk/jenkins
> -- The C compiler identification is AppleClang 8.1.0.8020041
> -- The CXX compiler identification is AppleClang 8.1.0.8020041
> -- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc
> -- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -- works
> -- Detecting C compiler ABI info
> -- Detecting C compiler ABI info - done
> -- Detecting C compile features
> -- Detecting C compile features - done
> -- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++
> -- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -- works
> -- Detecting CXX compiler ABI info
> -- Detecting CXX compiler ABI info - done
> -- Detecting CXX compile features
> -- Detecting CXX compile features - done
> -- CMake Install Prefix is /usr/local
> -- AIG architecture: x86_64
> -- Found PkgConfig: /usr/local/bin/pkg-config (found version "0.29.2") 
> -- NANOMSG LIBRARIES: nanomsg
> -- NANOMSG CFLAGS: -I/Users/user_name/src/try/azure-iot-gateway-sdk/install-deps/include
> -- NANOMSG LOCATION: /Users/user_name/src/try/azure-iot-gateway-sdk/install-deps/lib/libnanomsg.dylib
> -- Checking for one of the modules 'uuid'
> -- Configuring done
> -- Generating done
> CMake Warning:
>   Manually-specified variables were not used by the project:
>     run_e2e_tests
> -- Build files have been written to: /Users/user_name/src/try/azure-iot-gateway-sdk/build
> Scanning dependencies of target parson
> [  1%] Building C object core/deps/CMakeFiles/parson.dir/__/__/deps/parson/parson.c.o
> [  2%] Linking C static library libparson.a
> [  2%] Built target parson
> Scanning dependencies of target module_host_static
> Scanning dependencies of target gateway
> Scanning dependencies of target gateway_static
> [  5%] Building C object core/CMakeFiles/module_host_static.dir/adapters/dynamic_library_linux.c.o
> [  5%] Building C object core/CMakeFiles/module_host_static.dir/adapters/gb_library_linux.c.o
> [  7%] Building C object core/CMakeFiles/gateway_static.dir/adapters/dynamic_library_linux.c.o
> [  8%] Building C object core/CMakeFiles/gateway.dir/adapters/dynamic_library_linux.c.o
> [ 10%] Building C object core/CMakeFiles/gateway_static.dir/adapters/gb_library_linux.c.o
> [ 11%] Building C object core/CMakeFiles/module_host_static.dir/src/message.c.o
> [ 12%] Building C object core/CMakeFiles/gateway_static.dir/src/message.c.o
> [ 14%] Building C object core/CMakeFiles/gateway.dir/adapters/gb_library_linux.c.o
> [ 15%] Building C object core/CMakeFiles/gateway_static.dir/src/message_queue.c.o
> [ 17%] Building C object core/CMakeFiles/gateway.dir/src/message.c.o
> [ 18%] Building C object core/CMakeFiles/module_host_static.dir/src/message_queue.c.o
> [ 20%] Building C object core/CMakeFiles/gateway.dir/src/message_queue.c.o
> [ 21%] Building C object core/CMakeFiles/gateway_static.dir/src/module_loader.c.o
> [ 22%] Building C object core/CMakeFiles/module_host_static.dir/src/module_loader.c.o
> [ 24%] Building C object core/CMakeFiles/gateway.dir/src/module_loader.c.o
> [ 25%] Building C object core/CMakeFiles/gateway_static.dir/src/module_loaders/dynamic_loader.c.o
> [ 27%] Building C object core/CMakeFiles/module_host_static.dir/src/module_loaders/dynamic_loader.c.o
> [ 28%] Building C object core/CMakeFiles/gateway_static.dir/src/internal/event_system.c.o
> [ 30%] Building C object core/CMakeFiles/gateway.dir/src/module_loaders/dynamic_loader.c.o
> [ 31%] Building C object core/CMakeFiles/gateway.dir/src/internal/event_system.c.o
> [ 32%] Linking C static library libmodule_host_static.a
> [ 34%] Building C object core/CMakeFiles/gateway_static.dir/src/gateway_internal.c.o
> [ 35%] Building C object core/CMakeFiles/gateway.dir/src/gateway_internal.c.o
> [ 37%] Building C object core/CMakeFiles/gateway_static.dir/src/gateway.c.o
> [ 38%] Building C object core/CMakeFiles/gateway.dir/src/gateway.c.o
> [ 38%] Built target module_host_static
> [ 41%] Building C object core/CMakeFiles/gateway_static.dir/src/gateway_createfromjson.c.o
> [ 41%] Building C object core/CMakeFiles/gateway.dir/src/gateway_createfromjson.c.o
> [ 42%] Building C object core/CMakeFiles/gateway_static.dir/src/broker.c.o
> make[2]: *** No rule to make target `/usr/lib/x86_64-linux-gnu/libssl.so', needed by `core/libgateway.dylib'.  Stop.
> make[2]: *** Waiting for unfinished jobs....
> [ 44%] Building C object core/CMakeFiles/gateway.dir/src/broker.c.o
> [ 45%] Linking C static library libgateway_static.a
> make[1]: *** [core/CMakeFiles/gateway.dir/all] Error 2
> make[1]: *** Waiting for unfinished jobs....
> [ 45%] Built target gateway_static
> make: *** [all] Error 2

**Comment by damonbarry** [comment]
Your output looks a little different to me...I wonder if you're building an older version of the code. Are you sync'd up to the latest changes in the master branch? Can you reply with the output of `git describe`?

**Comment by kiranpradeep** [comment]
@damonbarry git describe gives `2017-04-27-31-g06dc9ad`. I have **minor** changes to avoid repeated clones and clean builds of libuv/java modules/remote mdoules. Copying git diff below.
```
machine_name:jenkins user_name$git diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 189aa19..d0229f0 100755
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -76,7 +76,7 @@ option(run_as_a_service "Flags that we have the goal of running gateway as a ser
 option(enable_dotnet_binding "set enable_dotnet_binding to ON to build dotnet binding host binaries (default is OFF)" OFF)
 option(enable_dotnet_core_binding "set enable_dotnet_core_binding to ON to build dotnet core binding host binaries (default is OFF)" OFF)
 option(enable_ble_module "set enable_ble_module to OFF to remove ble module from gateway build (default is ON)" ON)
-option(enable_native_remote_modules "Build the infrastructure required to support native remote modules (default is ON)" ON)
+option(enable_native_remote_modules "Build the infrastructure required to support native remote modules (default is ON)" OFF)
 option(enable_java_remote_modules "Build the infrastructure required to support java remote modules (default is OFF)" OFF)
 option(use_amqp "set use_amqp to ON if amqp is to be used, set to OFF to not use amqp" ON)
 option(use_http "set use_http to ON if http is to be used, set to OFF to not use http" ON)
diff --git a/jenkins/osx_c.sh b/jenkins/osx_c.sh
index e0b06bd..5bedd20 100755
--- a/jenkins/osx_c.sh
+++ b/jenkins/osx_c.sh
@@ -13,12 +13,12 @@ CORES=$(sysctl -n hw.ncpu)
 
 # Java binding
 pushd $build_root/bindings/java/gateway-java-binding
-mvn clean install
+#mvn clean install
 popd
 
 # Java proxy gateway
 pushd $build_root/proxy/gateway/java/gateway-remote-module
-mvn clean install
+#mvn clean install
 popd
 
 rm -rf $build_folder
@@ -27,13 +27,13 @@ pushd $build_folder
 
 cmake \
     -Ddependency_install_prefix:PATH=$build_root/install-deps \
-    -DOPENSSL_ROOT_DIR:PATH=/usr/local/opt/openssl \
+    -DOPENSSL_ROOT_DIR:PATH=/Users/user_name/base/openssl  \
     -Dbuild_cores=$CORES \
     -Denable_ble_module:BOOL=OFF \
-    -Denable_java_binding=ON \
-    -Denable_java_remote_modules:BOOL=ON \
-    -Drun_unittests:BOOL=ON \
-    -Drun_e2e_tests:BOOL=ON \
+    -Denable_java_binding=OFF \
+    -Denable_java_remote_modules:BOOL=OFF \
+    -Drun_unittests:BOOL=OFF \
+    -Drun_e2e_tests:BOOL=OFF \
     ..
 
 cmake --build . -- --jobs=$CORES
diff --git a/tools/build_libuv_macos.sh b/tools/build_libuv_macos.sh
index 3d65f2c..9de655c 100755
--- a/tools/build_libuv_macos.sh
+++ b/tools/build_libuv_macos.sh
@@ -10,15 +10,15 @@ build_root=$(cd "$(dirname "$0")/.." && pwd)
 build_root=$build_root/build_libuv
 
 # clear the libuv build folder so we have a fresh build
-rm -rf $build_root
-mkdir -p $build_root
+#rm -rf $build_root
+#mkdir -p $build_root
 
 # build libuv
 pushd $build_root
 
-git clone https://github.com/libuv/libuv.git
+#git clone https://github.com/libuv/libuv.git
 cd libuv
-git checkout -b v1.11.0 tags/v1.11.0
+#git checkout -b v1.11.0 tags/v1.11.0
 
 ./gyp_uv.py -f xcode
 xcodebuild -ARCHS="x86_64" -project uv.xcodeproj -configuration Release -target libuv
machine_name:jenkins user_name$ 
```

**Comment by damonbarry** [comment]
I don't have `libssl.so` on my Mac...I have a `dylib` under `/usr/local/opt/openssl/lib/`.

Where is the openssl library that you expect the build to find? (I assume it's under `/Users/user_name/base/openssl/`, but wanted to verify.) Also, do you have `pkgconfig/libssl.pc` under the same location? That's the file CMake will want to use (via pkgconfig) to figure out where openssl stuff lives.

**Comment by kiranpradeep** [comment]
I too don't have an so in my iMac :) Only .dylib installed via brew under /usr/local/opt/openssl/lib and custom built openssl under /Users/user_name/base/openssl. Each with it's own pkgconfig/libssl.pc. I had tried setting OPENSSL_ROOT_DIR to each directory with no difference.

Copying terminal below.
```
machine_name:lib user_name$ cd /usr/local/opt/openssl/lib/
machine_name:lib user_name$ ls -l
total 12232
drwxr-xr-x  14 user_name  admin      476 Jan 26 18:52 engines
-r--r--r--   1 user_name  admin  1995808 May 24 15:29 libcrypto.1.0.0.dylib
-r--r--r--   1 user_name  admin  3345384 Jan 26 18:52 libcrypto.a
lrwxr-xr-x   1 user_name  admin       21 Jan 26 18:52 libcrypto.dylib -> libcrypto.1.0.0.dylib
-r--r--r--   1 user_name  admin   373744 May 24 15:29 libssl.1.0.0.dylib
-r--r--r--   1 user_name  admin   528680 Jan 26 18:52 libssl.a
lrwxr-xr-x   1 user_name  admin       18 Jan 26 18:52 libssl.dylib -> libssl.1.0.0.dylib
drwxr-xr-x   5 user_name  admin      170 May 24 15:29 pkgconfig
machine_name:lib user_name$ 
machine_name:lib user_name$ 
machine_name:lib user_name$ cd /Users/user_name/base/openssl/lib/
machine_name:lib user_name$ ls -l
total 13240
drwxr-xr-x  14 user_name  TCS-MOBILITY\Domain Users      476 May 23 19:30 engines
-r-xr-xr-x   1 user_name  TCS-MOBILITY\Domain Users  2187184 May 23 19:30 libcrypto.1.0.0.dylib
-rw-r--r--   1 user_name  TCS-MOBILITY\Domain Users  3601044 May 23 19:30 libcrypto.a
lrwxr-xr-x   1 user_name  TCS-MOBILITY\Domain Users       21 May 23 19:30 libcrypto.dylib -> libcrypto.1.0.0.dylib
-r-xr-xr-x   1 user_name  TCS-MOBILITY\Domain Users   416224 May 23 19:30 libssl.1.0.0.dylib
-rw-r--r--   1 user_name  TCS-MOBILITY\Domain Users   560044 May 23 19:30 libssl.a
lrwxr-xr-x   1 user_name  TCS-MOBILITY\Domain Users       18 May 23 19:30 libssl.dylib -> libssl.1.0.0.dylib
drwxr-xr-x   5 user_name  TCS-MOBILITY\Domain Users      170 May 17 15:28 pkgconfig
machine_name:lib user_name$ 
machine_name:lib user_name$ 
machine_name:lib user_name$ ls -l pkgconfig/
total 24
-rw-r--r--  1 user_name  TCS-MOBILITY\Domain Users  305 May 23 19:30 libcrypto.pc
-rw-r--r--  1 user_name  TCS-MOBILITY\Domain Users  306 May 23 19:30 libssl.pc
-rw-r--r--  1 user_name  TCS-MOBILITY\Domain Users  244 May 23 19:30 openssl.pc
machine_name:lib user_name$ 
```

**Comment by damonbarry** [comment]
I see two options: either something goes wrong in our CMake scripts when openssl is not at the default location, or there's something wrong with your custom openssl installation.

I copied my default openssl tree (everything under usr/local/opt/openssl) to my home directory. Then I ran the CMake commands from the Jenkins script at my terminal, changing the value of OPENSSL_ROOT_DIR to /Users/damonb/openssl. When I did this, both project generation and build phases worked without error.

That would lead me to wonder if there's something wrong with your openssl files at /Users/user_name/base/openssl/.

Does the build work if you refer to the default install (/usr/local/opt/openssl)? What if you copy that tree to another location and reference it there...does the build work in that case?
